### ISSUE TITLE
Improve detection of libiconv and a C89 fix

### ISSUE THREAD

**Issue by pulkomandy** [jca02266/lha#16]
https://github.com/haikuports/haikuports/blob/master/app-arch/lha/patches/lha-1.14i.ac20191005.patchset

C89 fix: declaration after statement.
iconv: depending on how iconv is built, the symbol is named "iconv" or "libiconv". lha configure script detects libiconv in the library, but then checks again for the "iconv" symbol, and if not found, does not make use of iconv (but still links with it).

**Comment by ryandesign** [comment]
Thank you for showing how to detect iconv/libiconv properly. This explains a [build failure on macOS with Xcode 12](https://trac.macports.org/ticket/61535) as well. lha's code was intending to use iconv if available and fall back to private macOS functions if not, but since it did not detect the existence of libiconv properly it tried to use the private macOS functions even when libiconv was present. Since private functions are not in the OS headers, using them creates an implicit function declaration, which is not allowed anymore in Xcode 12. To fix the macOS case, these additional changes are needed:
```diff
--- src/header.c.orig	2019-10-04 22:58:55.000000000 -0500
+++ src/header.c	2020-12-06 23:31:25.000000000 -0600
@@ -1907,7 +1907,7 @@
 
 #if MULTIBYTE_FILENAME
 
-#if defined(__APPLE__) && !defined(HAVE_ICONV)  /* Added by Hiroto Sakai */
+#if defined(__APPLE__) && !(defined(HAVE_ICONV) || defined(HAVE_LIBICONV))  /* Added by Hiroto Sakai */
 
 #include <CoreFoundation/CFString.h>
 #include <CoreFoundation/CFStringEncodingExt.h>
@@ -2045,7 +2045,7 @@
 sjis_to_utf8(char *dst, const char *src, size_t dstsize)
 {
 #if defined(__APPLE__)
-#if !defined(HAVE_ICONV)
+#if !(defined(HAVE_ICONV) || defined(HAVE_LIBICONV))
   dst[0] = '\0';
   if (ConvertEncodingToUTF8(src, dst, dstsize,
                             kCFStringEncodingDOSJapanese,
@@ -2071,7 +2071,7 @@
 utf8_to_sjis(char *dst, const char *src, size_t dstsize)
 {
 #if defined(__APPLE__)
-#if !defined(HAVE_ICONV)
+#if !(defined(HAVE_ICONV) || defined(HAVE_LIBICONV))
   int srclen;
 
   dst[0] = '\0';
```

**Comment by ryandesign** [comment]
Actually I now see a previous report of libiconv issues in #13 and a fix merged in #14. So the fixes in this issue may be redundant, I'm not sure.

**Comment by pulkomandy** [comment]
This patch allows to use either the 'iconv' function, or the 'libiconv' function. Both exist on different operating systems. The current code (even after these other changes) I think still only handle only one case?

**Comment by jca02266** [comment]
Hmm, I would think that the `AC_SEARCH_LIBS(libiconv, [iconv])` add `-liconv` linker option.
But, the `AC_CHECK_FUNCS(iconv basename)` doesn't seem to find `iconv()` because it is actually `libiconv()`.

I think the `lha-1.14i.ac20191005.patchset` is correct.

I'm not sure if PR #14 was not enough.

**Comment by jca02266** [comment]
I tried recent LHa (28099f3) on Xcode12.2

> cd lha
> \# build GNU libiconv
> curl -O https://ftp.gnu.org/pub/gnu/libiconv/libiconv-1.16.tar.gz
> tar xvzf libiconv-1.16.tar.gz
> (cd libiconv-1.16 && mkdir build && cd build && ../configure && make check)
> \# build LHa
> (mkdir build && cd build && ../configure LDFLAGS='-L../libiconv-1.16/build/lib/.libs' CPPFLAGS='-I../libiconv-1.16/build/include' && make check)

libiconv was not found.
> grep ICONV config.h
> /* #undef HAVE_ICONV */
> \#define HAVE_ICONV_H 1

**Comment by jca02266** [comment]
I think LHa build sequence should work that:

1st search the iconv and it's know the UTF-8-MAC, select the libiconv
2nd search the iconv and it's unknow the UTF-8-MAC, select the libiconv (it's not Mac environ, Or it is select by user) 
3rd fallback the CoreFoundation on Mac
4th no use iconv

**Comment by jca02266** [comment]
Results of PR #19

# default (on Mac)
> (rm -rf build && mkdir build && cd build && ../configure && make check)
> 
> grep ICONV build/config.h
> > #define HAVE_ICONV 1
> > #define HAVE_ICONV_H 1
> > /* #undef HAVE_LIBICONV */
> > #define ICONV_HAS_UTF8MAC 1
> > #define USE_ICONV 1

# use GNU libiconv
> (rm -rf with-iconv && mkdir with-iconv && cd with-iconv && ../configure LDFLAGS='-L/opt/lib' CPPFLAGS='-I/opt/include' && make check)
> 
> grep ICONV build-with-gnulibiconv/config.h
> > /* #undef HAVE_ICONV */
> > #define HAVE_ICONV_H 1
> > #define HAVE_LIBICONV 1
> > /* #undef ICONV_HAS_UTF8MAC */
> > /* #undef USE_ICONV */

# use Core Foundation
> (rm -rf build-cf && mkdir build-cf && cd build-cf && ../configure --disable-iconv && make check)
> 
> grep ICONV build-cf/config.h
> > /* #undef HAVE_ICONV */
> > /* #undef HAVE_ICONV_H */
> > /* #undef HAVE_LIBICONV */
> > /* #undef ICONV_HAS_UTF8MAC */
> > /* #undef USE_ICONV */
> 
> otool -L build-cf/src/lha
> > build-cf/src/lha:
> > 	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1675.129.0)
> > 	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)

**Comment by jca02266** [comment]
Compile error problem on Xcode12 was fixed 959d248.

**Comment by jca02266** [comment]
Merged #19
